File: flatten-binary-tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void flatten(TreeNode root) {
        TreeNode head = new TreeNode(0);
        tail = head;
        makeFlat(root);
    }

    private void makeFlat(TreeNode root) {
        if (root == null) return;
        TreeNode left = root.left;
        TreeNode right = root.right;
        root.left = root.right = null;
        tail.right = root;
        tail = root;
        makeFlat(left);
        makeFlat(right);
    }

    private TreeNode tail;
}

File: flatten-binary-tree.cc
#include <cstdlib>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  void flatten(TreeNode *root) {
    TreeNode* head;
    TreeNode* tail;
    make_flatten(root, &head, &tail);
  }

  void make_flatten(TreeNode* root, TreeNode** head, TreeNode** tail) {
    if (!root) {
      *head = *tail = NULL;
      return;
    }
    *head = *tail = root;
    TreeNode* left_head, *left_tail, *right_head, *right_tail;
    make_flatten(root->left, &left_head, &left_tail);
    make_flatten(root->right, &right_head, &right_tail);
    bool has_left = root->left;
    bool has_right = root->right;
    root->left = NULL;
    if (has_left) {
      (*tail)->right = left_head;
      *tail = left_tail;
    }
    if (has_right) {
      (*tail)->right = right_head;
      *tail = right_tail;
    }
  }
};
